Docs: Update the Getting Started section

robmurgai opened the pull request:
By Rob Murgai 

#### Type of change

- Documentation update

#### Description

The "Getting Started" Docs have not been updated in a bit and we would like to bring the instructions back in synch with the latest release, and make the Section easier to follow for folks new to Fabric .

Today, the Getting Started looks like this:

**Getting Started**
- Prerequisites
- Install Samples, Binaries, and Docker Images
- Using the Fabric test network
- Hyperledger Fabric smart contract (chaincode) APIs
- Hyperledger Fabric application SDKs
- Hyperledger Fabric CA


This PR breaks the section in two, and updates the individual pages as needed.

**Getting Started - Install**
- Prerequisites
- Install Fabric Samples                             (Formerly - Install Samples, Binaries, and Docker Images)
- Install Fabric Application SDKs              (Formerly - Hyperledger Fabric application SDKs)
- Install Fabric Contract APIs                    (Formerly - Hyperledger Fabric smart contract (chaincode) APIs)

**Getting Started - Run Fabric**
- Running A Test network                          (Formerly - Using the Fabric test network)
- Running A Fabric Application                 (Writing Your First Application - Sample Basic Asset Transfer)


#### Additional details

- The Prerequisites and Fabric install instructions have been broken into a happy path on top and a support section at the bottom. This provides a default for most folks to get started with Fabric quickly and scroll down for details, troubleshooting, as needed.  The install instructions have also been broken down by environments (MAC, Linux, Win) to make it easier for those new to fabric, to get going quickly.
- The Samples install instructions were updated with an optional set of instructions to download the Fabric repo.  The idea is to build on the developer's existing interest in Fabric and give them the option to have the Fabric code base on their machines so they can delve into the code, check out implementation, etc.  The Samples install instructions were maintained on top for continuity.
- "Getting Started - Run Fabric" section has been organized to start the user with the tutorial on test network and follow it with the tutorial on "Basic Asset Transfer Sample". The intent of this inclusion was to enable new users to get comfortable with both - the basics of a fabric network and a fabric application.  
- The TOC link to Test Network Tutorial was updated from "Using the Fabric Test Network" to "Running A Test Network".  For a new user who just got done installing Samples, it feels more intuitive to "run" a Fabric Network, followed by "run" a Fabric Application on that test network, rather than "use" a test network.  For consistency with the rest of the documentation, the original title of the tutorial remains unchanged to build the understanding that the test network can and does get used as the basis for our tutorials and samples, and is a tool to be leveraged.


#### Related issues

srderson commented:
@christo4ferris Thanks, this is more clear. LGTM

robmurgai committed:
Docs: Update Getting Started section with install commands and add links to asset transfer basic.

This update will bring the install portions of Getting Started in synch with the latest release amd makes the process easier to follow for folsk new to Fabric. 

Getting Started broken into two sections:

1. Getting Started Install
       Install instructions for Prerequisite softwar
       Instal instructions for Fabric Samples
       Install links for SDKS and Chaincode APIs
2. Getting Started Run Fabric
       Run a Test Network.
       Run a Fabric Application.

Signed-off-by: Rob Murgai <smurgai@gmail.com>

nikhil550 commented:
some other commits seem to have snuck inside this documentation change

robmurgai commented:
> some other commits seem to have snuck inside this documentation change

Thanks @nikhil550 fixed.  Thanks @sykesm for the help with rebase.

pamandrejko commented:
Looking over your PR I was also curious about this existing topic for “developers” https://hyperledger-fabric.readthedocs.io/en/latest/dev-setup/devenv.html?highlight=development%20env

It seems to overlap with your new content and was curious if you planned to leave this as is or differentiate that content somehow.

I’m still unclear on who/what the new topics are targeted for? I see 3 types of “tasks”
SysAdmins who want to setup and run a Fabric network.
Developers who want to use Fabric, write client app or smart contract.
Developers who want to contribute to Fabric.

robmurgai commented:
Thanks for your feedback Pam. 

> Looking over your PR I was also curious about this existing topic for “developers” https://hyperledger-fabric.readthedocs.io/en/latest/dev-setup/devenv.html?highlight=development%20env
> 
> It seems to overlap with your new content and was curious if you planned to leave this as is or differentiate that content somehow.

Good question. My thought was to create one area with install instructions for everything, so for newbies it would be easier to go back to if someone had issues and for maintainers its easier to update things in one place.

So I reviewed the "Setting up the development environment" document and merged the Fabric install instructions with Fabric Samples install instructions.  And to be honest, it sounds like it would be a lot but its a handful of things that get added - like in pre-reqs you add jq and softHSM, and in Fabric & Samples install you add a couple of make commands.

Plus personally, it was a lot more satisfying to install Fabric and run the make command and see stuff come back successful.

And from a  flow perspective, my thought was as newbies start playing with samples and want to graduate to developing their own applications, setting up their own networks, etc, they can just start and not worry about installs, etc.

> I’m still unclear on who/what the new topics are targeted for? I see 3 types of “tasks”
> SysAdmins who want to setup and run a Fabric network.
> Developers who want to use Fabric, write client app or smart contract.
> Developers who want to contribute to Fabric.

I am going to assume you were referring to the 4 Install pages when you said "topics":

1. Prerequisites
2. Install Fabric (& Samples)
3. Install Fabric Application SDKs
4. Install Fabric Contract APIs

The topics are targeted for anyone new to Fabric.  I was thinking at this stage, people are just "Getting Started", and everyone is still looking to understand Fabric and how it works.  So, I wasn't really thinking so much about instructions as what would a Network Admin need vs what would a Developer need.  I was thinking more along the lines of what do I need to have to understand Fabric and build my depth. For me that be play with samples like asset transfer and test network which explain things pretty well, and look at the Fabric and Sample code base to follow the logic of the explanation in those tutorials.  

So the instructions are set up to meet that goal of having Fabric and Samples on your machine.

Once people got comfortable with Fabric, I figured they would probably pick specific tutorials and reference documentation to continue building their skills in their areas of focus whether its Network Administration or app development, so the personas would come more into play there.  And, if in the future they wanted to start contributing to Fabric, they could do that as well, it would be an almost organic next step.

PS: I did make one exception to persona specific install steps, where I said if you are not looking to build applications and just want to understand how Fabric networks work, you could skip installing SDK/APIs

nikhil550 commented on docs/source/index.rst:
```diff
@@ -19,7 +19,7 @@ modularity and versatility for a broad set of industry use cases.
    whatsnew
    key_concepts
    getting_started
```
The developing applications topic is important and we should leave it in the table of contents

robmurgai replied:
Agreed.  I had moved it into the Getting Started - Run Fabric Section, but it should stay as part of the main TOC. I got the same feedback from others as well, "Developing Applications" deserves its own section.  I have moved it back.

nikhil550 commented on docs/source/test_network.md:
```diff
@@ -1,4 +1,4 @@
-# Using the Fabric test network
+# Running A Fabric network
```
I would leave title. The topic is not about running a Fabric network generally, its about using the test network specifically. There are also many links to it as well.

robmurgai replied:
Agreed, It's a good point @nikhil550. The idea behind updating the title from "Using the Fabric Test Network" to "Running A Fabric Network" was to see it from a new user perspective who at this stage would not know enough to understand that the test network is a lot more versatile than a tutorial on running the network and that it can and does get used as the basis for our tutorials and samples.  For a new user who just got done installing Samples, it feels more intuitive to run a Fabric Network and then (if needed) run a Fabric Application on that test network.  It also creates a pattern:  (1) Run a Fabric Network (2) Run a Fabric Application.

That being said, I think it misses the point that test network is a tool that can be used for a lot of things.  One of the recommendation I got was to merge the ideas, I have updated the TOC link as "Running a Test Network".  However, the original title of the Tutorial is unchanged - "Using the Fabric Test Network".

nikhil550 commented on docs/source/write_first_app.rst:
```diff
@@ -1,4 +1,4 @@
-Writing Your First Application
+Running A Fabric Application
```
I am more sympathetic to this title change. Running a Fabric application is more descriptive than writing your first app. But I would check and update the links to the article as well

robmurgai replied:
Agreed.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,188 @@
+
+> **Fabric:** Install Fabric in `$HOME/go/src/github.com/<your_github_userid>` directory.  This is a Golang Community recommendation for Go projects and this location will also enable local directories to be shared with Docker containers
+
+>> **Fork** the Upstream Repo. Navigate to [(https://github.com/hyperledger/fabric)](https://github.com/hyperledger/fabric) and fork the fabric repository using the fork button in the top-right corner. 
```
This seems to mix the instructions for Fabric developers and for those who are getting started. I think the contributors section is a better place for the small number of users who are going to contribute to the project

robmurgai replied:
Agreed - we need to make sure its the right set of install work for a new developer and we also want to keep it consistent with how we have been doing things to date. This was also discussed in the documentation calls and taking everyone's feedback, I have redone the page - where we start with Fabric Samples and give people the option to download Fabric if they want but the install/make/run instructions have been removed.  

The idea behind giving developers (By developer I mean anyone interested in Fabric - not someone looking to Contribute necessarily) the option to download the Fabric repo was to build on the their interest in Fabric - they are here because they want to understand and use Fabric - so lets give them the option to have Fabric on their machines so they can delve into the code, check out implementation, etc.  But let's not complicate it with make or running integration tests instructions.  And yes, let's keep the Samples instructions on top for continuity.

nikhil550 reviewed (commented):
I think we should narrow the scope of this PR. The pre-requisites is a large improvement. But I do not think we have enough data to overhaul the entire getting started journey, especially since the personnas for Fabric are not as well defined in practice as they are for other technologies. I think we should do incremental PR's to the journey, get more feedback, and see where the improvements lead. The new asset transfer series and the test network both had very positive impacts on the documentation that way. It would also make things easier for users who may already be familiar with the existing docs.

pamandrejko commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,30 @@
-install HyperLedger Fabric. While we work on developing real installers for the
-Fabric binaries, we provide a script that will :doc:`install` to your system.
-The script also will download the Docker images to your local registry.
+The install process for Fabric has 4 steps and each step lines up loosely with your Blockchain Application Stack.
```
spell out numbers < 9. so 4 should be four

robmurgai replied:
Done.

pamandrejko commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,30 @@
-After you have downloaded the Fabric Samples and Docker images to your local
-machine, you can get started working with Fabric with the
-:doc:`test_network` tutorial.
+.. image:: ./getting_started_image2.png
```
Really love the new pyramid. There are five layers but they don't match up to your four bullets below it. Perhaps you should add a 5th bullet for application or call it out separately.

robmurgai replied:
Called it our separately.

pamandrejko commented on docs/source/getting_started_run_fabric.rst:
```diff
@@ -0,0 +1,20 @@
+   write_first_app
+
+
+Now that you have installed Fabric and Samples, you can start running Fabric.  (If you have not installed all the required software, please follow the instructions on :doc:`getting_started` first).
```
In general we don't use "please" in technical documentation.

robmurgai replied:
So not how I was raised - but deleted.

pamandrejko commented on docs/source/install.md:
````diff
@@ -0,0 +1,137 @@
+  ```
+
+
+* GOPATH: By default, you do not need to set the GOPATH environment variable, it is assumed to be $HOME/go on Unix systems and %USERPROFILE%\go on Windows. If however, you would like to use a different location for fabric-samples install, you may set GOPATH to point to your specific go workspace. For example on macOS 
````
Thanks for clarifying this.

pamandrejko commented on docs/source/getting_started_run_fabric.rst:
```diff
@@ -0,0 +1,20 @@
+  * If you are interested in understanding how Fabric Networks work, we recommend starting with our :doc:`Running a Test Network <test_network>` tutorial.  Its the next topic in this section.
+  * If you are interested in developing blockchain applications on top of a Fabric Network, you are welcome to start with our :doc:`write_first_app` tutorial.  Its the second topic in this section. However, we do recommend comming back to :doc:`Running a Test Network <test_network>` tutorial for completeness.
+
+Both tutorials will link to deeper explanations in :doc:`key_concepts`.  We leave it to you to jump off to the concept or ton continue with the tutorial.  
```
typo ton continue

robmurgai replied:
fixed

pamandrejko commented on docs/source/test_network.md:
```diff
@@ -376,7 +376,7 @@ local machine, you can use the tutorials to start developing your own solution:
 
```
The top of the Test Network tutorial under Before you begin still refers to to a topic titled  "`Installed the Samples, Binaries and Docker Images.` which I believe is now renamed. 


Before you can run the test network, you need to clone the fabric-samples repository and download the Fabric images. Make sure that you have installed the Prerequisites and Installed the Samples, Binaries and Docker Images.

robmurgai replied:
updated.

pamandrejko commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,232 @@
+> jq-1.6
+>```
+
+> **SoftHSM**
````
Should we say that installing SoftHSM is (Optional). Only required if you plan to play around with HSM?

I think what I am missing in this section is,  are ALL these prereqs required or if you only want to run the test network for example, is there a subset that are required? Is there a baseline of prereqs required and then if you plan to do xxx then you also need to install yyy?

robmurgai replied:
Mostly, I think we need all but the SoftHSM prereqs.  We have 7 in total:
**Git, cURL, Docker, GoLang** are needed for the install.  
**Wget**  I remember having to update it on my machine when I was installing Fabric Samples, its needed to download multiple files in your system and I think(??) it's used indirectly by one of the other commands.  
**JQ** is used to update channel configuration when use use `confixtxlator` in our tutorials, like [here ](https://hyperledger-fabric.readthedocs.io/en/latest/channel_update_tutorial.html?highlight=jq#convert-the-configuration-to-json-and-trim-it-down).
**SoftHSM** - You are correct, you only need SoftHSM if you want to use HSM.  I have updated the section with "SoftHSM is Optional - Please install if you plan to play use HSM in your application"

pamandrejko commented on docs/source/deploy_chaincode.md:
```diff
@@ -32,10 +32,24 @@ The `createChannel` command creates a channel named ``mychannel`` with two chann
-- [Step two: Install the chaincode package](#install-the-chaincode-package)
-- [Step three: Approve a chaincode definition](#approve-a-chaincode-definition)
-- [Step four: Committing the chaincode definition to the channel](#committing-the-chaincode-definition-to-the-channel)
+- [Deploying a smart contract to a channel](#deploying-a-smart-contract-to-a-channel)
```
Since this was a tutorial I think Nik's intent here was to list the overall steps for chaincode deployment so people understand the process rather than provide a tutorial toc. Maybe you are thinking of something different here?

nikhil550 replied:
It also lined up to the chaincode lifecycle key concept doc, so the concept and the tutorial was alligned

robmurgai replied:
Not sure why this change got introduced, it seemed to be an automatic TOC type update.  Anyway, I have manually changed it back to the original 4 steps.

pamandrejko reviewed (commented):
Coming along nicely Rob. Thanks again for putting all of this together. A few nits to consider.

pamandrejko commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,232 @@
+>```
+
+> **SoftHSM**
+> Installing SoftHSM is Optional - Please install if you plan to play use HSM in your application.  Install the latest version of [softhsm](https://wiki.opendnssec.org/display/SoftHSMDOCS/SoftHSM+Documentation+v2) if it is not already installed.
````
typo
Please install if you plan to play use HSM. And in general we don't use 'Please' in technical content.

robmurgai replied:
Done.

pamandrejko commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,232 @@
+>- We have had users report errors while running Fabric-Samples with `Docker Desktop 2.4.0.0(48506)` when the `gRPC FUSE for file sharing` option is checked.  Please uncheck this option in your Docker Preferences to continue using `osxfs for file sharing`.
+>- The `curl` command that comes with Git and Docker Toolbox is old and does not handle properly the redirect used in [Getting Started](https://hyperledger-fabric.readthedocs.io/en/latest/getting_started.html). Please make sure you use a newer version from the [cURL downloads page](https://curl.haxx.se/download.html)
+>- These prerequisites are recommended for Fabric users. If you are a Fabric developer, please refer to the instructions for [Setting up the development environment](https://hyperledger-fabric.readthedocs.io/en/latest/dev-setup/devenv.html).
+>- Configure SoftHSM: Fabric uses the PKCS #11 API to interact with hardware security modules (HSMs) that store cryptographic information and perform cryptographic computations and part of our unit tests require  cryptographic tokens to test this functionality.  SoftHSM is an implementation of a cryptographic store that's accessible through a PKCS#11 interface and can be used for test environment ...
```
Is there a some action that users need to take because of this or is this just for their awareness? Is this a if you plan to use SoftHSM kind of thing only? I guess I am trying to understand why is it important to tell them this?

robmurgai replied:
Yeah, its an answer to "What is SoftHSM", it's not needed in this section.  Removed.

pamandrejko commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
+
+**Fabric-Samples:**<br>
+We think the best way to understand something is to use it yourself.  To help you use Fabric, we have created a simple Fabric Network called `test-net` and a set of sample applications that demonstrate its core capabilities. We have also precompiled `Fabric CLI tool binaries` and `Fabric Docker Images` which will be donwloaded to your environment, to get you going.
```
typo: donwloaded

robmurgai replied:
Fixed.

pamandrejko commented on docs/source/sdk_chaincode.md:
```diff
@@ -0,0 +1,53 @@
+Install Fabric Application SDKs
+-----------------------------------
+
+**PS:** For Network Administrators interested in running and managing a Fabric Network only, you may skip installing the following SDKs and APIs.
```
Instead of **PS**, we've used an **Audience** header instead.
See https://hyperledger-fabric.readthedocs.io/en/latest/tutorial/commercial_paper.html 
https://hyperledger-fabric.readthedocs.io/en/latest/tutorial/commercial_paper.html for an example of that pattern which I think could work nicely here.

robmurgai replied:
Updated.

pamandrejko commented on docs/source/test_network.md:
```diff
@@ -20,9 +20,9 @@ current master branch, it is possible that you will encounter errors.
-Before you can run the test network, you need to clone the `fabric-samples`
-repository and download the Fabric images. Make sure that you have installed
-the [Prerequisites](prereqs.html) and [Installed the Samples, Binaries and Docker Images](install.html).
+Before you can run the test network, you need to install Fabric Samples in your
+environment. Follow the instructions on [getting_started](getting_started.html)
+to install the required software.
```
There is a reference in  test_network.md file to the "Windows extras" section of the prereqs which no longer exists.
I would suggest changing it from
This was most likely caused by not setting
   ``core.autocrlf`` to ``false`` in your Git configuration (see
    [Windows extras](prereqs.html#windows-extras)).
 to
 This was most likely caused by not setting
   ``core.autocrlf`` to ``false`` in your Git configuration (see
    [Windows](prereqs.html#windows)).

robmurgai replied:
Done.

pamandrejko reviewed (commented):
Wow Rob. This has come a long way. It nicely follows progressive disclosure, is logical, well-organized, and concise. I think it is pretty close to ready. I found a few minor suggestions and also I checked to see what links would break. In addition to the ones already noted, we should also update this one:

In commerical_paper.md there is a reference to [Download the Samples, Binaries and Docker Images](../install.html). It should be changed to [Install Fabric Samples](../install.html).

Thank you for taking this on. I know it was a massive effort probably larger than you initially anticipated.

nikhil550 commented on docs/source/index.rst:
```diff
@@ -19,6 +19,7 @@ modularity and versatility for a broad set of industry use cases.
    whatsnew
    key_concepts
    getting_started
+   getting_started_run_fabric
```
I like the new section. I feel like we should have always done it that way.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+# Install Fabric Samples
```
When I first read through this, I kept looking for "install Fabric", ie how do I get it on my machine. I wonder if we should rename this - Install Fabric and samples (we can omit the binaries since no-one at this point knows that those are)

denyeart replied:
agreed

robmurgai replied:
Done.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+<br>
+<br>
+
+#### Mac
```
It looks like the commands are the same for all of them, do we need three different sections. May be just say "all platforms"?

robmurgai replied:
Answered the comment down below for brevity.

nikhil550 commented on docs/source/install.md:
````diff
@@ -0,0 +1,137 @@
+> 
+>> **Install**  Install Fabric-Samples in `$HOME/go/src/github.com/<your_github_userid>` directory.  This is a Golang Community recommendation for Go projects and this location will also enable local directories to be shared with Docker containers
+>> ```shell
+>> $ mkdir -p $HOME/go/src/github.com/<your_github_userid>
````
Is it necessary to clone the repo's separately, since the bootstrap.sh command already does that? or are we changing the flow? (cloning the repo does make things more deliberate, I can image people running the command from anywhere accidentally)

is the `go/src/` path necessary? That sound required for developers? also, the samples contains more scripts and javascript than any go code.

robmurgai replied:
Answered the comment down below for brevity.

nikhil550 commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,231 @@
+# Prerequisites
+
+Let's start with Prerequisites:
+
```
I would include a set of links to each platform section at the top here, similar to some of our tutorials.

robmurgai replied:
@nikhil550 I didn't see a specific one @nikhil550, but if you point me to one, I can keep them consistent.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
+
+**Fabric-Samples:**<br>
```
If we consolidate the platform sections, we can make these the sections of this article:
- Fabric docker images and Fabric samples
- Fabric source code

robmurgai replied:
@nikhil550 I can, but the slashes would be wrong.  I mean it would be obvious to windows users, but still forward vs backward slashes would be incorrect.

 Also, doesn't Windows use different variables for $HOME ?

>> $ mkdir -p $HOME/go/src/github.com/<your_github_userid>
>> $ cd $HOME/go/src/github.com/<your_github_userid>

However, as you said, if we remove the instructions to install the files in this directory altogether, then collapsing makes complete sense.  And technically you are right, we don't need to install Samples here, but I had the Fabric code base as an optional install in this page which made following the same directory structure for both repos $HOME/go/src/github.com/<your_github_userid> more efficient.  

But I have since removed the optional Fabric Install section, so . . . we could just delete it and collapse the three environments into one section.

@denyeart @pamandrejko any issues with swapping this section:

> Install Install Fabric-Samples in $HOME/go/src/github.com/<your_github_userid> directory.  This is a Golang Community recommendation for Go projects and this location will also enable local directories to be shared with Docker containers
> 
> $ mkdir -p $HOME/go/src/github.com/<your_github_userid>
> $ cd $HOME/go/src/github.com/<your_github_userid>


With

> Enter the directory where you want to install Fabric in a terminal window and enter the following cURL command.

denyeart replied:
I've collapsed the platform installs into a single download section, and added a few platform-specific considerations at the end.

denyeart commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,31 @@
-Hyperledger Fabric smart contract (chaincode) APIs
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+* Prerequisite software - is the base layer of all the needed software.
+* Fabric Samples - will have the executables to run a Fabric Network along with sample code.
```
Let's call it `Fabric and Fabric Samples` (the first time I read the bullet list I said 'what about Fabric itself?')

robmurgai replied:
Done

denyeart commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,31 @@
+* Prerequisite software - is the base layer of all the needed software.
+* Fabric Samples - will have the executables to run a Fabric Network along with sample code.
+* Contract APIs - to develop smart contracts executed on a Fabric Network.
+* Application SDKs - to develop your blockchain application.
```
You don't really 'install' contract APIs and application SDKs. You just reference them as dependencies in your chaincode and applications. So having them in the install list is a bit misleading.

robmurgai replied:
@denyeart @nikhil550 @pamandrejko technically true.

I think the contract APIs and application SDKs do have some installation like the node sdks require node to be downloaded I think but those are minimal and it makes more sense for those install instructions to be in the Tutorials themselves.  For instance, the simple asset transfer tutorial has a [Before you Begin](https://hyperledger-fabric.readthedocs.io/en/release-2.3/write_first_app.html#before-you-begin) section that instructs on node and npm install.

With that thinking, I took a stab at modifying the install section (getting_started.html) into 2 Steps:
1. Prereqs
2. Binaries, images and Samples.

Unless, y'all have issues, I am going to add the contract APIs and application SDK introduction to that page  (getting_started.html) below the triangle image.

Please let me me know if you think it's a better fit elsewhere or if we should remove the introduction to contract APIs and application SDKs altogether.

denyeart replied:
In my last pushes, I've added the Fabric Contract APIs and Application SDKs page to the "Getting Started - Install" submenu.

denyeart commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,31 @@
-After you have downloaded the Fabric Samples and Docker images to your local
-machine, you can get started working with Fabric with the
-:doc:`test_network` tutorial.
+.. image:: ./getting_started_image2.png
```
I'm loving the new pyramid!

denyeart commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,31 @@
 
-Hyperledger Fabric offers a number of APIs to support developing smart contracts (chaincode)
-in various programming languages. Smart contract APIs are available for Go, Node.js, and Java:
+The Application layer on top is the blockchain application you will develop to run in a Fabric Network.
```
```suggestion
The Application layer on top is the blockchain application you will develop to run on top of a Fabric Network.
```

denyeart commented on docs/source/getting_started.rst:
```diff
@@ -1,64 +1,31 @@
-that you may choose to use to generate the certificates and key material
-to configure and manage identity in your blockchain network. However, any CA
-that can generate ECDSA certificates may be used.
+This section will help you install the required software and the next section will help you run a network and an application on it.
```
When a top level page has child pages, it is best to have them in a bullet list with links.

denyeart commented on docs/source/getting_started_run_fabric.rst:
```diff
@@ -0,0 +1,20 @@
+
+  * If you are interested in understanding how Fabric Networks work, we recommend starting with our :doc:`Running a Test Network <test_network>` tutorial.  Its the next topic in this section.
+  * If you are interested in developing blockchain applications on top of a Fabric Network, you are welcome to start with our :doc:`write_first_app` tutorial.  Its the second topic in this section. However, we do recommend comming back to :doc:`Running a Test Network <test_network>` tutorial for completeness.
+
```
Instead of saying 'next topic' and 'second topic', just let the bullets speak for themselves, e.g.
- Running a test network (link) - short description
- Running a Fabric application (link) - short description

denyeart commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
+
+**Fabric-Samples:**<br>
+We think the best way to understand something is to use it yourself.  To help you use Fabric, we have created a simple Fabric Network called `test-net` and a set of sample applications that demonstrate its core capabilities. We have also precompiled `Fabric CLI tool binaries` and `Fabric Docker Images` which will be downloaded to your environment, to get you going.
```
```suggestion
We think the best way to understand something is to use it yourself.  To help you use Fabric, we have created a simple Fabric Network called `test-network` using docker compose, and a set of sample applications that demonstrate its core capabilities. We have also precompiled `Fabric CLI tool binaries` and `Fabric Docker Images` which will be downloaded to your environment, to get you going.
```

robmurgai replied:
done.

denyeart commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+We think the best way to understand something is to use it yourself.  To help you use Fabric, we have created a simple Fabric Network called `test-net` and a set of sample applications that demonstrate its core capabilities. We have also precompiled `Fabric CLI tool binaries` and `Fabric Docker Images` which will be downloaded to your environment, to get you going.
+
+**Optional Fabric Download:**<br>
+We are huge proponents of reviewing actual code whereever possible to get a solid understanding of technical concepts. If you feel the same way, we have added instructions to help you download the Fabric code base on to your system.  This is a personal preference, you do not need this code to run any of the samples.
```
Most people kicking the tires will not want to look at the Fabric code, I would simply include a link to "Setting up the development environment" somewhere at the bottom, instead of talking about this here at the top.

robmurgai replied:
Done.

denyeart commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,231 @@
+> Release-Date: 2019-03-27
+> ```
+
+> **wget**
````
wget is not needed anymore.

denyeart commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,231 @@
+> GNU Wget 1.20.3 built on darwin19.0.0.
+> ```
+
+> **Docker**
````
This is Docker Desktop right?
Note that there are issues with Docker Desktop v3.x, so we'll need to say latest v2.x instead of latest.

robmurgai replied:
I thought the issue was with Docker Desktop 2.4.0.0 when "gRPC FUSE for file sharing" option was checked?  I have Docker Desktop 3.0.4 with that option unchecked and it seems to be okay with the test network?

denyeart commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,231 @@
+> ```
+
+
+> **GoLang** 
````
Only needed if developing Fabric right? I'd defer this to "Setting up the dev environment" topic.

robmurgai replied:
@denyeart Are you referring to docker-compose?

denyeart commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,231 @@
+> go1.14.6 darwin/amd64
+>```
+
+> **JQ**
````
Only needed if doing config transactions, not needed for the two basic tutorials that you recommend starting with.

robmurgai replied:
@denyeart true - if the install section is just for the two tutorials then we could add the JQ install instructions in the Tutorials that reference config transactions, here for instance: https://hyperledger-fabric.readthedocs.io/en/latest/channel_update_tutorial.html#setup-the-environment

But, I thought the install section was to enable the users to be able to use any tutorial without more installs?

denyeart commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,231 @@
+>```
+
+> **SoftHSM**
+> Installing SoftHSM is Optional - Please install if you plan to play use HSM in your application.  Install the latest version of [softhsm](https://wiki.opendnssec.org/display/SoftHSMDOCS/SoftHSM+Documentation+v2) if it is not already installed.
````
SoftHSM not needed by most users kicking the tires, just defer to Setting up the dev env topic.

denyeart commented on docs/source/sdk_chaincode.md:
```diff
@@ -0,0 +1,53 @@
+Install Fabric Application SDKs
```
You don't 'install' the SDKs, you just reference them in your application.

robmurgai replied:
Removed the word Install

denyeart commented on docs/source/sdk_chaincode.md:
```diff
@@ -0,0 +1,53 @@
+model delivered in Hyperledger Fabric v1.4.
+
+
+Install Fabric Contract APIs
```
You don't 'install' the contract APIs, you just reference them in your chaincode.

robmurgai replied:
Removed the work Install.

denyeart commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+> You have completed installing Fabric samples, binaries and docker images to your system.
+<br>
+
+> **Optional Fabric Download**<br>
```
I wouldn't put this in the main install page at all, just link to "Setting up the development environment".

denyeart commented on docs/source/install.md:
```diff
@@ -0,0 +1,137 @@
+
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
+
+**Fabric-Samples:**<br>
```
Instead of `Fabric-Samples`, I'd call the section `Download Fabric images, binaries, and samples`

denyeart reviewed (commented):
Nice work, overall I love the new structure:
- Getting Started - Install
- Getting Started - Run Fabric
  - Running a Test Network
  - Running a Fabric Application

The new structure makes it much more obvious what the first steps are.

My largest concerns below fall into two categories
- Make it clear that we are not just installing fabric samples, but fabric images, binaries, and samples
- We should exclude all content related to setting up the development environment (cloning fabric, installing developer prereqs), we already have a decent "Setting up the development environment" topic that we can link to.

pamandrejko commented on docs/source/sdk_chaincode.md:
```diff
@@ -0,0 +1,53 @@
+
+> -   [Python SDK](https://github.com/hyperledger/fabric-sdk-py).
+> -   [Go SDK](https://github.com/hyperledger/fabric-sdk-go).
+
```
The Go SDK is now v1.0.0 so I think it can be moved to the list above. Would this be the readme?
https://github.com/hyperledger/fabric-sdk-go/tree/v1.0.0

denyeart replied:
I think the link for docs would be:
https://pkg.go.dev/github.com/hyperledger/fabric-sdk-go

robmurgai replied:
Done.

denyeart committed:
Update Prereq, Install, and SDK docs

Update prereq and install doc for formatting and minor edits.
Update SDK doc to match latest doc updates in master.

Signed-off-by: David Enyeart <enyeart@us.ibm.com>

denyeart commented:
I've resolved the merge conflicts and re-pushed.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,120 @@
+# Install Fabric and Fabric Samples
+
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
```
I have a few comments on this page.
1. I though we were moving away from the gopath, and that it was not required.
2. I thought that the install script cloned the samples for you, and that doing that separately was not required.
3. The steps in this doc seem the be the same for all platforms. 
My thinking would be to to revert this doc to something closer to the current [Install](https://hyperledger-fabric.readthedocs.io/en/latest/install.html) topic and merge, since the rest is a large improvement.

denyeart replied:
Thanks @nikhil550 , good comments, I've addressed in the latest push.

joealewine commented on docs/source/getting_started.rst:
```diff
@@ -1,66 +1,29 @@
-install HyperLedger Fabric. While we work on developing real installers for the
-Fabric binaries, we provide a script that will :doc:`install` to your system.
-The script also will download the Docker images to your local registry.
+The Fabric application stack has 5 layers:
```
Fabric is an "application stack"? Are we comfortable with that term?

Also at a minimum this should be "five".

denyeart replied:
Switched to "five". Open to suggestions on application stack.

joealewine commented on docs/source/getting_started.rst:
```diff
@@ -1,66 +1,29 @@
-Getting Started
-===============
+Getting Started - Install
```
These hyphens everywhere aren't necessarily the best. They should be em dashes or even better colons

denyeart replied:
Switched to colons.

joealewine commented on docs/source/getting_started.rst:
```diff
@@ -1,66 +1,29 @@
+* :doc:`Fabric and Fabric Samples <install>` - the Fabric executables to run a Fabric Network along with sample code.
+* :doc:`Contract APIs <sdk_chaincode>` - to develop smart contracts executed on a Fabric Network.
+* :doc:`Application SDKs <sdk_chaincode>` - to develop your blockchain application.
+* The Application - your blockchain application will utilize the Application SDKs to call smart contracts running on a Fabric Network.
```
Why is "samples" and "network" capitalized here? And again, these hyphens would be better as colons.

denyeart replied:
done.

joealewine commented on docs/source/getting_started_run_fabric.rst:
```diff
@@ -0,0 +1,20 @@
+   write_first_app
+
+
+Now that you have installed Fabric and Samples, you can start running Fabric.  (If you have not installed all the required software, follow the instructions on :doc:`getting_started` first).
```
I feel like a broken record, but there is no reason to capitalize "samples" here. It is not a proper noun. Also, I can't see a reason this second sentence is inside a parenthesis. It's a perfectly reasonable sentence to have here.

denyeart replied:
done.

joealewine commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+
+Please install the [Prerequisites](./prereqs.html) before following these install instructions.
+
+We think the best way to understand something is to use it yourself.  To help you use Fabric, we have created a simple Fabric Network called `test-network` using docker compose, and a set of sample applications that demonstrate its core capabilities.
```
Again, "Fabric network" is fine. This is not the proper name of a particular network.

denyeart replied:
done.

joealewine commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+
+The cURL command in the instructions below sets up your environment so that you can run the 'test-network'. Specifically, it performs the following steps:
+* Clones the [hyperledger/fabric-samples](https://github.com/hyperledger/fabric-samples) repository.
+* Downloads the latest Hyperledger Fabric docker images and tags them as `latest`
```
"Docker", however, is a brand name. And should be capitalized.

denyeart replied:
done.

joealewine commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+The cURL command in the instructions below sets up your environment so that you can run the 'test-network'. Specifically, it performs the following steps:
+* Clones the [hyperledger/fabric-samples](https://github.com/hyperledger/fabric-samples) repository.
+* Downloads the latest Hyperledger Fabric docker images and tags them as `latest`
+* Downloads the following platform-specific Hyperledger Fabric CLI tool binaries and config files into the fabric-samples `/bin` and `/config` directories. These binaries will help you interact with the 'test-network'.
```
In this context, `fabric-samples` should be inside backtics.

denyeart replied:
done.

joealewine commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+
+## Download Fabric samples, docker images, and binaries.
+
+Download fabric-samples to the `$HOME/go/src/github.com/<your_github_userid>` directory.  This is a Golang Community recommendation for Go projects. If you are using a different directory or Windows, see the [Notes](https://hyperledger-fabric.readthedocs.io/en/latest/install.html#notes) below.
```
And again, re: fabric-samples

denyeart replied:
done.

joealewine commented on docs/source/install.md:
````diff
@@ -0,0 +1,85 @@
+curl -sSL https://bit.ly/2ysbOFE | bash -s -- -h
+```
+
+To download a specific release, pass a version identifier for Fabric and Fabric-CA docker images. The command below demonstrates how to download the latest production releases - `Fabric v2.3.1` and `Fabric CA v1.4.9` 
````
The Fabric CA generally calls itself "Fabric CA", not "Fabric-CA: https://hyperledger-fabric-ca.readthedocs.io/en/latest/deployguide/ca-deploy.html

Also "Docker" again here.

denyeart replied:
done.

joealewine commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+## Troubleshooting
+
+* If you get an error running the curl command
+  * You may have too old a version of curl that does not handle redirects or an unsupported environment. Please make sure you use a newer version from the [cURL downloads page](https://curl.haxx.se/download.html)
```
cURL

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,222 @@
+# Prerequisites
+
+The following prerequisites are required to run a docker-based Fabric test network on your local machine.
```
Docker

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,222 @@
+sudo apt-get -y install docker-compose
+```
+
+Once installed, confirm that the latest versions of both docker and docker-compose executables were installed.
````
Docker

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,222 @@
+docker-compose version 1.27.2, build 18f557f9
+```
+
+Make sure the docker daemon is running.
````
Docker

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,222 @@
+sudo systemctl start docker
+```
+
+Optional: If you want the docker daemon to start when the system starts, use the following:
````
Docker

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
````diff
@@ -0,0 +1,222 @@
+sudo systemctl enable docker
+```
+
+Add your user to the docker group.
````
Sigh.

Docker.

denyeart replied:
done.

joealewine commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,222 @@
+
+### Docker
+
+Install the latest version of [Docker](https://docs.docker.com/get-docker/) if it is not already installed.
```
Oh thank God.

joealewine commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,222 @@
+
+Install the latest version of [Docker](https://docs.docker.com/get-docker/) if it is not already installed.
+
+Once installed, confirm the latest versions of both docker and docker-compose executables were installed.
```
Welp. 

Docker.

joealewine commented on docs/source/prereqs.md:
```diff
@@ -0,0 +1,222 @@
+
+## **Notes**
+
+- The `curl` command that comes with Git and Docker Toolbox is old and does not handle properly the redirect used in [Getting Started](https://hyperledger-fabric.readthedocs.io/en/latest/getting_started.html). Please make sure you use a newer version from the [cURL downloads page](https://curl.haxx.se/download.html)
```
cURL

Be consistent.

denyeart replied:
done.

nikhil550 commented on docs/source/install.md:
```diff
@@ -0,0 +1,85 @@
+The cURL command in the instructions below sets up your environment so that you can run the 'test-network'. Specifically, it performs the following steps:
+* Clones the [hyperledger/fabric-samples](https://github.com/hyperledger/fabric-samples) repository.
+* Downloads the latest Hyperledger Fabric Docker images and tags them as `latest`
+* Downloads the following platform-specific Hyperledger Fabric CLI tool binaries and config files into the `fabric-samples` `/bin` and `/config` directories. These binaries will help you interact with the 'test-network'.
```
Should test network be in backticks to match style at the top. I do not have any preference (other docs just call it the Fabric test network), but we should be consistant.

denyeart replied:
I've made it Fabric test network consistently.

nikhil550 reviewed (commented):
One small nit, but this is looking close to being ready to merge.

nikhil550 reviewed (approved):
LGTM. I will leave unmerged in case Joe has any additional comments.

joealewine commented:
The application stack thing is weird to me but seems like Dave got everything I saw -- which was superficial stuff anyway.

denyeart commented:
Ok, thanks for the reviews @nikhil550 and @joealewine , and for the initial content @robmurgai . Great teamwork, merging!
